### PR TITLE
test: add e2e for self-update in linux

### PR DESCRIPTION
**Pull request by vjripoll**
Adds Linux E2E self-update test

  ## Test Flow

  1. Install AC 1.13.0 using New Relic CLI recipe
  2. Push test version (0.900.xxx) to local OCI registry
  3. Configure self-update + OpAMP endpoint
  4. Verify initial version (1.13.0)
  5. Send remote config with new version
  6. Verify self-update completes successfully

### CODE REVIEW

File: test/e2e-runner/src/common/cert.rs [hunk 1]
```diff
@@ -35,26 +35,38 @@ impl SelfSignedCert {
...
+        "subjectAltName=DNS:localhost,IP:127.0.0.1",
+    ]);
+
+    #[cfg(target_family = "unix")]
+    cmd.args([
+        "-addext",
+        "basicConstraints=critical,CA:FALSE",
+        "-addext",
+        "keyUsage=critical,digitalSignature,keyEncipherment",
+    ]);
```
Comment (gsanchezgavier) [inline test/e2e-runner/src/common/cert.rs]: would this work on win too? just to unify the implementation

File: test/e2e-runner/src/common/cert.rs [hunk 2]
```diff
@@ -77,6 +89,37 @@ fn add_to_root_store(cert_path: &std::path::Path) {
...
+        .expect("failed to run update-ca-certificates");
+
+    assert!(
+        output.status.success(),
+        "update-ca-certificates failed: {}",
+        String::from_utf8_lossy(&output.stderr)
+    );
+}
+
+impl Drop for SelfSignedCert {
```
Comment (gsanchezgavier) [inline test/e2e-runner/src/common/cert.rs]: is this needed for these ephermeral tests?

File: test/e2e-runner/src/linux/install.rs [hunk 3]
```diff
@@ -107,6 +107,44 @@ curl -Ls https://download.newrelic.com/install/newrelic-cli/scripts/install.sh |
...
+  NEW_RELIC_LICENSE_KEY={} \
+  NEW_RELIC_API_KEY={} \
+  NEW_RELIC_ACCOUNT_ID={} \
+  NEW_RELIC_AUTH_PROVISIONED_CLIENT_ID={} \
+  NEW_RELIC_AUTH_PRIVATE_KEY_PATH={} \
+  NEW_RELIC_REGION={} \
+  NR_CLI_FLEET_ID={} \
+  NEW_RELIC_AGENT_CONTROL_FLEET_ENABLED={} \
+  NEW_RELIC_AGENT_CONTROL=true \
+  NEW_RELIC_AGENT_VERSION=1.13.0 \
```
Comment (gsanchezgavier) [inline test/e2e-runner/src/linux/install.rs]: this version was hardcoded for the test, you should be able to remove it so it always fetch the latest

File: test/e2e-runner/src/linux/scenarios/self_update.rs [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+                Err(format!("expected version {new_version}, got {reported_version}").into())
+            }
+        },
+    );
+    info!(version = new_version, "AC version updated successfully");
+
+    info!("Self-update test completed successfully");
+}
+
+fn find_ac_instance_id(server: &FakeServer) -> TestResult<InstanceID> {
```
Comment (gsanchezgavier) [inline test/e2e-runner/src/linux/scenarios/self_update.rs]: perhaps makes sense to reuse this from a common place ?

File: CHANGELOG.md [hunk 5]
```diff
@@ -11,6 +11,9 @@ Unreleased section should follow [Release Toolkit](https://github.com/newrelic/r
 ### 🚀 Enhancements
 - Added support for JP endpoints
 
+### 🧪 Testing
```
Comment (gsanchezgavier) [inline CHANGELOG.md]: oh i didn't know there was this kind in the release toolkit

Comment (gsanchezgavier) [reply CHANGELOG.md]: oh no , whatch out , this is working because of a bug in the RT i think 🤔 , i just tried to run the `make rt-update-changelog` with only an entry like this and 💥

File: test/e2e-runner/src/linux/service.rs [hunk 6]
```diff
@@ -2,10 +2,73 @@ use tracing::info;
...
+/// Restarts a service using systemctl and waits for it to reach the expected status
+pub fn restart_service_and_wait(service_name: &str, expected_status: &str) {
+    info!(service = service_name, "Restarting service");
+    let cmd = format!("systemctl restart {service_name}");
+    let _ = exec_bash_command(&cmd)
+        .unwrap_or_else(|err| panic!("could not restart the '{service_name}' service: {err}"));
+
+    // Wait for service to reach expected status
+    use std::thread::sleep;
+    use std::time::Duration;
```
Comment (sigilioso) [inline test/e2e-runner/src/linux/service.rs]: Nit: is there any particular reason ti import these inside the function?

File: test/e2e-runner/src/linux/service.rs [hunk 7]
```diff
@@ -2,10 +2,73 @@ use tracing::info;
...
+
+    for i in 0..30 {
```
Comment (sigilioso) [inline test/e2e-runner/src/linux/service.rs:25]: I was wondering if we could simplify this with the retry helper, but maybe not if we'd like show the info log every 5th iteration

File: test/e2e-runner/src/linux/service.rs [hunk 8]
```diff
@@ -2,10 +2,73 @@ use tracing::info;
...
+            info!(
+                service = service_name,
+                current_status = current_status,
+                expected_status = expected_status,
+                "Waiting for service to reach expected status"
+            );
+        }
+        sleep(Duration::from_secs(1));
+    }
+    let final_status = get_service_status(service_name);
```
Comment (sigilioso) [inline test/e2e-runner/src/linux/service.rs:47]: This is really corner-case: what happens if we get the expected status right after finishing the loop?

File: test/e2e-runner/src/linux/scenarios/self_update.rs [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+        "Verified initial AC version before self-update"
+    );
+
+    let new_version = pushed_package.reference.tag().unwrap();
+    assert_ne!(
+        initial_version, new_version,
+        "initial and new version must differ for self-update to be meaningful"
+    );
+
+    let update_config = format!("version: \"{new_version}\"\nagents: {{}}");
```
Comment (sigilioso) [inline test/e2e-runner/src/linux/scenarios/self_update.rs]: nit: I'd use the the `r#""#` strings to improve readability if possible, something similar to:

```rust
    let update_config = format!(r#"
version: "{new_version}"
agents: {{}}
    "#);
```

**Review by sigilioso — COMMENTED** [review]
Considering how complex is the scenario these tests are verifying they are still pretty easy to follow 👏 

As discussed here https://github.com/newrelic/newrelic-agent-control/pull/2490#issuecomment-4405088150 we really need to also test the other way around when possible (I think it is already planned).

I've left some nits, let me know your thoughts.

File: test/e2e-runner/src/common/cert.rs (hunk 1 above)
Comment (vjripoll) [reply test/e2e-runner/src/common/cert.rs]: yes true simplfied

File: test/e2e-runner/src/common/cert.rs (hunk 2 above)
Comment (vjripoll) [reply test/e2e-runner/src/common/cert.rs]: yep, for these tests are not needed because as you said

File: test/e2e-runner/src/linux/scenarios/self_update.rs (hunk 4 above)
Comment (vjripoll) [reply test/e2e-runner/src/linux/scenarios/self_update.rs]: moved to the fake opamp to have encapsulated the logic there

File: CHANGELOG.md (hunk 5 above)
Comment (vjripoll) [reply CHANGELOG.md]: yep, i dind't read the readme of the changelog 👍

File: test/e2e-runner/src/linux/service.rs (hunk 6 above)
Comment (vjripoll) [reply test/e2e-runner/src/linux/service.rs]: moved to the top

File: test/e2e-runner/src/linux/service.rs (hunk 7 above)
Comment (vjripoll) [reply test/e2e-runner/src/linux/service.rs:25]: what you prefer ?i think is hepful to have it

File: test/e2e-runner/src/linux/scenarios/self_update.rs (hunk 9 above)
Comment (vjripoll) [reply test/e2e-runner/src/linux/scenarios/self_update.rs]: ❤️

File: test/e2e-runner/src/linux/service.rs (hunk 7 above)
Comment (sigilioso) [reply test/e2e-runner/src/linux/service.rs:25]: Well, if we used the retry helper, we could always print the message on every iteration... I actually don't have a strong opinion

File: test/e2e-runner/src/linux/scenarios/self_update.rs (hunk 4 above)
Comment (gsanchezgavier) [reply test/e2e-runner/src/linux/scenarios/self_update.rs]: have you pushed the changes i don't see it moved

File: CHANGELOG.md [hunk 10]
```diff
@@ -10,6 +10,7 @@ Unreleased section should follow [Release Toolkit](https://github.com/newrelic/r
 
 ### 🚀 Enhancements
 - Added support for JP endpoints
+- Added e2e test for self-update functionality on Linux
```
Comment (gsanchezgavier) [inline CHANGELOG.md]: not sure if this belong to new features , IMO i would skip it fro the changelog

**Review by gsanchezgavier — APPROVED** [review]
thanks! skip-changelog label should work for the CI check to pass
